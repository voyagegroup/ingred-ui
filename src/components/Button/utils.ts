import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
