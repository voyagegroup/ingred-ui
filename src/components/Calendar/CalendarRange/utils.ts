import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
