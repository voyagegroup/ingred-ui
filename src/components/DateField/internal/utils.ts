export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
