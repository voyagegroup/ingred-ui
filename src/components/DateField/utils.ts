import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
