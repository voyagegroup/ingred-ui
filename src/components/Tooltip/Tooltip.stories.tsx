import React from "react";
import Spacer from "../Spacer";
import Flex from "../Flex";
import Tooltip from "./Tooltip";
import { TooltipProps } from "..";
import { Story } from "@storybook/react/types-6-0";

export default {
  title: "Components/Data Display/Tooltip",
  component: Tooltip,
};

export const Example: Story<TooltipProps> = (args) => {
  return <Tooltip {...args} />;
};

Example.args = {
  children: <span>Hover me!!</span>,
  content: "Sample text",
};

export const DesignSamples = () => {
  return (
    <>
      <Flex display="flex" justifyContent="center">
        {["top-start", "top", "top-end"].map((position) => (
          <Spacer key={position} p={8}>
            <Tooltip
              open={true}
              content={position.toUpperCase()}
              positionPriority={[position as any]}
            >
              <span>{position.toUpperCase()}</span>
            </Tooltip>
          </Spacer>
        ))}
      </Flex>
      <Flex display="flex" justifyContent="space-between" alignItems="center">
        <Spacer pl={12}>
          {["left-start", "left", "left-end"].map((position) => (
            <Spacer key={position} p={4}>
              <Tooltip
                open={true}
                content={position.toUpperCase()}
                positionPriority={[position as any]}
              >
                <span>{position.toUpperCase()}</span>
              </Tooltip>
            </Spacer>
          ))}
        </Spacer>
        <Spacer pr={12}>
          {["right-start", "right", "right-end"].map((position) => (
            <Spacer key={position} p={4}>
              <Tooltip
                open={true}
                content={position.toUpperCase()}
                positionPriority={[position as any]}
              >
                <span>{position.toUpperCase()}</span>
              </Tooltip>
            </Spacer>
          ))}
        </Spacer>
      </Flex>
      <Flex display="flex" justifyContent="center">
        {["bottom-start", "bottom", "bottom-end"].map((position) => (
          <Spacer key={position} p={8}>
            <Tooltip
              open={true}
              content={position.toUpperCase()}
              positionPriority={[position as any]}
            >
              <span>{position.toUpperCase()}</span>
            </Tooltip>
          </Spacer>
        ))}
      </Flex>
    </>
  );
};
