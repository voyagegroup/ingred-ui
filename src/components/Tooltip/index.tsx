export { default, TooltipProps } from "./Tooltip";
