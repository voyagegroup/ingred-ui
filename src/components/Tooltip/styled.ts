import styled, { DefaultTheme } from "styled-components";
import * as PopperJS from "@popperjs/core";

const getArrowPosition = (
  theme: DefaultTheme,
  placements: PopperJS.Placement[],
  currentPlacement?: PopperJS.Placement,
) =>
  currentPlacement && placements.includes(currentPlacement)
    ? `${theme.spacing * 1.5}px`
    : "auto";

export const Arrow = styled.div<{ placement?: PopperJS.Placement }>`
  position: absolute;
  top: ${({ theme, placement }) =>
    getArrowPosition(theme, ["left-start", "right-start"], placement)};
  bottom: ${({ theme, placement }) =>
    getArrowPosition(theme, ["left-end", "right-end"], placement)};
  left: ${({ theme, placement }) =>
    getArrowPosition(theme, ["top-start", "bottom-start"], placement)};
  right: ${({ theme, placement }) =>
    getArrowPosition(theme, ["top-end", "bottom-end"], placement)};

  z-index: -1;
  &:after {
    content: " ";
    position: absolute;
    width: 8px;
    height: 8px;
    background-color: ${({ theme }) => theme.palette.black};
  }
`;

export const Tooltip = styled.div<{ width?: string }>`
  display: flex;
  align-items: center;
  align-content: center;
  width: ${({ width }) => width || "auto"};
  padding: ${({ theme }) => `${theme.spacing * 0.75}px ${theme.spacing}px`};
  border-radius: ${({ theme }) => theme.radius}px;
  color: ${({ theme }) => theme.palette.white};
  background-color: ${({ theme }) => theme.palette.black};
  font-weight: bold;
  font-size: 12px;
  z-index: ${({ theme }) => theme.depth.tooltip};

  &[data-popper-placement^="top"] > ${Arrow} {
    bottom: 0;
    &:after {
      top: -5px;
      transform: translateX(-50%) rotate(45deg)
        skew(calc((30deg) / 2), calc((30deg) / 2));
    }
  }

  &[data-popper-placement^="bottom"] > ${Arrow} {
    top: 0;
    &:after {
      bottom: -5px;
      transform: translateX(-50%) rotate(45deg)
        skew(calc((30deg) / 2), calc((30deg) / 2));
    }
  }

  &[data-popper-placement^="right"] > ${Arrow} {
    left: 0;
    &:after {
      right: -5px;
      transform: translateY(-50%) rotate(-45deg)
        skew(calc((30deg) / 2), calc((30deg) / 2));
    }
  }

  &[data-popper-placement^="left"] > ${Arrow} {
    right: 0;
    &:after {
      left: -5px;
      transform: translateY(-50%) rotate(-45deg)
        skew(calc((30deg) / 2), calc((30deg) / 2));
    }
  }
`;
