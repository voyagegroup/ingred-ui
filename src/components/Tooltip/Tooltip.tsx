import * as React from "react";
import * as Styled from "./styled";
import Fade from "../Fade";
import { Placement } from "@floating-ui/core";
import { useMergeRefs } from "../../hooks/useMergeRefs";
import { CSSTransitionProps } from "../../utils/reactTransitionGroup";
import {
  autoPlacement,
  flip,
  useFloating,
  offset as floatingOffset,
  shift,
  arrow,
  autoUpdate,
  useRole,
  useInteractions,
  FloatingPortal,
  useHover,
  FloatingArrow,
  FloatingDelayGroup,
} from "@floating-ui/react";
import { useTheme } from "../../themes";
import { AutoPlacement, usePlacement } from "../../hooks/usePlacement";

export type TooltipProps = React.ComponentPropsWithoutRef<"div"> & {
  content: React.ReactNode;
  open?: boolean;
  disableHoverListener?: boolean;
  enterDelay?: number;
  leaveDelay?: number;
  /**
   * Define priority of position. Please check [this](https://floating-ui.com/docs/tutorial#placements).
   */
  positionPriority?: (Placement | AutoPlacement)[];
  offset?: [number, number];
  width?: string;
  disabled?: boolean;
  /**
   * props of [Fade](/?path=/docs/components-utils-fade)
   */
  fadeProps?: CSSTransitionProps;
  children: React.ComponentElement<HTMLElement, any>;
};

const Tooltip = React.forwardRef<HTMLDivElement, TooltipProps>(function Tooltip(
  {
    content,
    open: openProp = false,
    disableHoverListener = false,
    enterDelay = 0,
    leaveDelay = 0,
    positionPriority = ["top"],
    offset = [0, 10],
    width,
    disabled = false,
    fadeProps,
    children,
    ...rest
  },
  ref,
) {
  const { placements, isAuto } = usePlacement(positionPriority);
  const arrowRef = React.useRef(null);
  const [open, setOpen] = React.useState<boolean>(false);
  const theme = useTheme();

  const {
    x,
    y,
    placement,
    refs: floatingRef,
    strategy,
    context,
  } = useFloating({
    placement: placements[0],
    open: open,
    onOpenChange: setOpen,
    middleware: [
      positionPriority.length > 0 && !isAuto
        ? flip({
            fallbackPlacements: placements,
          })
        : autoPlacement(),
      floatingOffset({
        mainAxis: offset[1],
        crossAxis: offset[0],
      }),
      shift({
        mainAxis: false,
      }),
      arrow({
        element: arrowRef,
      }),
    ],
    whileElementsMounted: autoUpdate,
  });

  const hover = useHover(context, {
    enabled: !disableHoverListener,
    move: false,
    delay: {
      open: enterDelay,
      close: leaveDelay,
    },
  });
  const role = useRole(context, { role: "tooltip" });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    hover,
    role,
  ]);

  const childrenProps = {
    ...children.props,
    ...getReferenceProps(),
    ref: useMergeRefs(floatingRef.setReference, children.ref),
  };

  const refs = useMergeRefs<HTMLDivElement>(ref, floatingRef.setFloating);

  return (
    <>
      {React.cloneElement(children, childrenProps)}
      {!disabled && (
        <FloatingDelayGroup delay={{ open: 10000, close: 200 }}>
          <FloatingPortal>
            <Fade
              in={open || openProp}
              unmountOnExit={true}
              mountOnEnter={true}
              {...fadeProps}
            >
              <Styled.Tooltip
                ref={refs}
                style={{
                  position: strategy,
                  top: y,
                  left: x,
                }}
                {...rest}
                {...getFloatingProps()}
                width={width}
              >
                {content}
                <FloatingArrow
                  ref={arrowRef}
                  fill={theme.palette.black}
                  context={context}
                  width={8}
                  height={8}
                  staticOffset={(() => {
                    if (
                      [
                        "top-start",
                        "top-end",
                        "bottom-start",
                        "bottom-end",
                      ].includes(placement)
                    ) {
                      return "10%";
                    }
                    if (
                      [
                        "bottom-start",
                        "bottom-end",
                        "left-start",
                        "left-end",
                      ].includes(placement)
                    )
                      return "30%";
                    return undefined;
                  })()}
                />
              </Styled.Tooltip>
            </Fade>
          </FloatingPortal>
        </FloatingDelayGroup>
      )}
    </>
  );
});

export default Tooltip;
