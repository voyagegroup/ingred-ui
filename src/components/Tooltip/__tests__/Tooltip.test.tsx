import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";
import Tooltip from "../Tooltip";

jest.mock("react-dom", () => {
  const original = jest.requireActual("react-dom");
  return {
    ...original,
    createPortal: (node: any) => node,
  };
});

describe("Tooltip component testing", () => {
  afterEach(cleanup);

  test("Tooltip", () => {
    const { asFragment } = renderWithThemeProvider(
      <Tooltip content="tooltip text" open={true}>
        <div>text</div>
      </Tooltip>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
