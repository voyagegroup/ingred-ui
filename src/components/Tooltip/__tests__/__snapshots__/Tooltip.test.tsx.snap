// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component testing Tooltip 1`] = `
<DocumentFragment>
  <div>
    text
  </div>
  <div
    class="sc-dkPtRN jMueqr fade-transition-appear fade-transition-appear-active"
  >
    <div
      class="sc-gsDKAQ chwUAn"
      style="position: absolute; left: 0px; top: 0px;"
    >
      tooltip text
      <div
        class="sc-bdvvtL bIjFMV"
        data-popper-arrow="true"
      />
    </div>
  </div>
</DocumentFragment>
`;
